### PR TITLE
Fix resume not working in some cases

### LINKED ISSUE #1017: Resume playback not working.
Hi, I am having a problem with resume. It doesn't work ever, all shows and movies start from the beginning even though I choose "resume from XX:xx" Files saved on local HDD and played directly resume correctly. I also have a less common issue where kodi seems to lock up when i exit playback. This issue occurred again when I ran the test of resume to produce the following log.
I have not experienced either of these issues prior t using PKC. 
Thanks.

https://pastebin.com/4wzLgDhx

### PR DESCRIPTION
**Pull request by croneter**
- Fixes #1017

### CODE REVIEW

**Comment by peacey** [comment]
Hey croneter, thanks for working on this. I tested this commit and it fixes resume but with one small change. Your if/else resume logic in _playback_init will never get to the last else block to check force_resume because api.resume_point() will always be > 0.0 if an item is partially played. So this code is bringing up a context menu even though force_resume is True. Checking if force_resume is True before opening the context menu fixes this logic. This was my attempt at fixing it:

`if app.PLAYSTATE.resume_playback is not None:
        # Kodi already displayed dialog for user whether he'd like to resume
        LOG.debug('Using app.PLAYSTATE.resume_playback for resume: %s',
                  app.PLAYSTATE.resume_playback)
        resume = app.PLAYSTATE.resume_playback
    elif api.resume_point():
        if force_resume is not None:
            LOG.debug('Using force_resume %s', force_resume)
            resume = force_resume
        else:
            # Since Kodi won't ask if user wants to resume playback -
            # we need to ask ourselves
            LOG.debug('Showing PKC resume dialog')
            resume = resume_dialog(int(api.resume_point()))
            if resume is None:
                LOG.debug('User aborted resume dialog')
                return
    else:
        resume = False`

Also, I was wondering whether the code that brings up the resume dialog is redundant (including the app.PLAYSTATE.resume_playback stuff)? Kodi will always pass the resume parameter in the plugin call, whether resume:true or resume:false. If the default select action in the Kodi settings is set to "Play" or "Choose", then when the user selects a video, Kodi will open a dialog that asks the user to resume or play from beginning, and if they select resume then Kodi will pass resume:true, and if they select play from beginning then Kodi will pass resume:false to the plugin. If the default select action is set to "Resume" instead, then Kodi will not open a dialog but will set "resume:true" in the plugin call. If the user uses the context menu and selects resume, then Kodi passes resume:true as well. If the skin uses a PlayMedia call, it also passes the resume:true/resume:false parameter. Basically, there is no scenario where the resume parameter is not passed to the PKC plugin by Kodi. Are there any other cases I'm missing where Kodi doesn't bring up a dialog or doesn't pass the resume parameter? Maybe in an older version of Kodi?

**Comment by croneter** [comment]
@peacey Thats exactly what I am working on right now - the PKC logic for resume seems completely redundant. The issue I am having now though is that I want to run through init playback first, then conclude playback on a second pass (for trailers). BUT: Kodi is passing resume:false for the second pass for some reason (I am trying to drop appending `resume=1` to the addon path for the second run

**Comment by peacey** [comment]
I did a lot of debugging but couldn't get Kodi to ever pass `resume:true` on the second pass. I don't believe `xbmc.player.play` (which kicks off the second pass), supports setting the `resume:true` parameter. I also tried to use the JSON RPC API with Player.Open because it does have a resume parameter, but resume:true wasn't passed even when I supplied the parameter as boolean, time, or percentage. So I think there is something broken in Kodi with the resume passing in these functions whether it's the python API or JSON RPC.

However, I did figure out that you don't need to append `resume=1` for the second pass. The first pass already checks for resume in `_playback_init` and then passes the offset to `_threaded_playback` which then seeks to the offset once the video starts playing. I just made a small change to allow seeking to a zero/None offset in `_threaded_playback` and I disabled the `_set_resume` call for second passes in `_conclude_playback` because it seemed redundant. I still kept it enabled for other scenarios that only have first passes like playlist_playback.

Also, there was a small bug with play_xml where the offset parameter is received in milliseconds from plex_companion, but then passed to _threaded_playback without converting to seconds first, which was screwing up my casting "resume/play from beginning" actions in some cases.

All the resume functionality (resume & play from beginning) now work with these changes, in both Kodi and casting scenarios. Not sure if these changes will introduce any new issues though. 

Here's my changes in a patch file: 
[pkc-fix-resume-2.patch.txt](https://github.com/croneter/PlexKodiConnect/files/3775705/pkc-fix-resume-2.patch.txt)

**Comment by peacey** [comment]
Hey croneter, I found a couple more bugs with resume in some edge cases.

1. There is a race condition that causes the JSON Player.Seek call (js.seek_to) to fail and return an error in some cases (specifically in the case where you're switching videos very quickly). It would return an error like this:

`{u'message': u'Failed to execute method.', u'code': -32100}`

It could be because even after you confirm there is a player id, the player can disappear before/while calling Player.Seek and then reappear after because of the transition from queue to player. Whatever the reason, I think a solution would be to try to run js.seek_to multiple times until it succeeds to be safe (e.g. 10 tries max with 0.1 s interval). That fixes the problem for me in the edge case of switching videos really quickly. The seek method rarely fails the first time, but on the off-chance it fails, it has never failed the second time in my tests. 

2. Sometimes if you play the same video twice, but not consecutively after each other, the resume point might not update on the API, so it resumes on the incorrect position next play. You solve this in _set_resume by looking up the resume point from the DB instead of using item.resume. Now that you moved resume to _playback_init, you should add the DB lookup to there too before calling threaded_playback so the correct offset is passed. That fixed this problem for me. 

3. The previous commit already fixed this last problem (by not using _set_resume), but I figured this out while debugging so I thought I would tell you. If you play the same video multiple times consecutively after each other with no other videos in between, Kodi resumes/plays from beginning properly without you doing anything depending on what the user chose (the plugin doesn't activate on the subsequent plays in this case because Kodi is just replaying the video in the Player). However, if you set the listitem property StartPercent, it forces Kodi to start the video from the same point each time and breaks the ability to resume/play from beginning in the case when the same video is played twice after each other (e.g.: user chooses to resume first time, but then closes the video and decides to play it from beginning). So it's best not to set the StartPercent property for GUI/companion playback so it doesn't interfere with anything. 

Here's another patch file with my attempt at the above fixes: 
[pkc-fix-resume-3.patch.txt](https://github.com/croneter/PlexKodiConnect/files/3782449/pkc-fix-resume-3.patch.txt)

**Comment by croneter** [comment]
Hah, I could kiss you, thanks so much for the info about `StartPercent` and the other listitem parameters. Simply dropping them altogether fixed everything, as far as I can tell. 

Does it work now?

**Comment by peacey** [comment]
Lol, thanks croneter. Glad I could help! So with the latest changes, all cases seem to be working but I had to change a couple minor issues in the commit first.

1. One of your later commits reversed the PKC casting commit that added /1000 to convert milliseconds to seconds in play_xml. I had to add it in again. 

2. In threaded_playback, you use the check `if offset:`, which is False if offset is 0. So play from beginning wasn't working until I changed the condition to allow a zero offset. Also, sometimes a None (not 0) offset is passed to threaded_playback (e.g.: from play_xml when user tries PKC casting with play from beginning). I think you should convert a None offset to 0 before the if condition in threaded_playback so that it always seeks to 0 if offset is either 0 or None. Doing something like this makes it work again:

```
offset = offset or 0
if offset >= 0:
```

3. Just an observation, but I realized you're no longer using the resume parameter passed to _conclude_playback (since you disabled _set_resume entirely). Is it necessary to pass it as a parameter to that function if you no longer use it? Or is there some other need for it?

With those changes, everything seems to be working now :-)

**Comment by croneter** [comment]
Great. I'll fix that. Can you comment code like I did above? Would be easier :wink:

Playlists are Kodi playlists. Add e.g. A couple of movies to a playqueue, then safe as playlist, then navigate to video playlists...

File: resources/lib/playback.py [hunk 1]
```diff
@@ -581,21 +558,38 @@ def play_xml(playqueue, xml, offset=None, start_plex_id=None):
...
+    to 0 (starting position) in case Kodi wants to resume but we want to start
+    over.
+
+    offset: resume position in seconds [int/float]
     """
+    LOG.debug('threaded_playback with startpos %s, offset %s',
+              startpos, offset)
     app.APP.player.play(kodi_playlist, None, False, startpos)
-    if offset and offset != '0':
+    if offset:
```
[the remark this answers is not in the document]
Comment (peacey) [reply resources/lib/playback.py]:
```suggestion
    offset = offset or 0
    if offset >= 0:
```

File: resources/lib/playback.py [hunk 2]
```diff
@@ -564,8 +541,8 @@ def play_xml(playqueue, xml, offset=None, start_plex_id=None):
     playqueue.selectedItemID
     """
     offset = int(offset) if offset else None
```
Comment (peacey) [inline resources/lib/playback.py]: Convert milliseconds to seconds
```suggestion
    offset = int(offset)/1000 if offset else None
```

**Comment by croneter** [comment]
Still need to test Kodi playlist playback, otherwise we should be good

File: resources/lib/playback.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (peacey) [reply resources/lib/playback.py]: Are you going to commit this change? I see you resolved the conversation but the changes weren't committed.

**Comment by peacey** [comment]
So I tested playlist playback quite a bit. The good thing is that Kodi seems to be able to handle resume/play from beginning without you explicitly programming it. Probably because the resume:true/resume:false parameter is passed and Kodi knows how to handle it (the plugin doesn't get called a second time with playlist playback, so the resume:true parameter isn't overridden by resume:false the second time around, which screws up Kodi's ability to resume without us programming it explicitly). So I don't believe we have to pass the resume parameter to _playlist_playback.

It's good resume works, but I did find some bugs while I was debugging. Here are my findings:

1. Resume/play from beginning works fine when user clicks on playlist item and Kodi asks to resume/play from beginning. It also works fine when default select action is set to resume and user clicks on item. 

2. Bringing up the context menu on a playlist item, there is no "Play from Beginning" button. This is probably because of the way the items are added to playlist (type unknown and all) and Kodi is confused about what to do. Couldn't think of any solution for this...

3. In the context menu, there is a "Resume from xx" option, but clicking it results in an error and doesn't play anything. The error is

`WARNING: Playlist Player: ListItem type must be audio or video, use ListItem::setInfo to specify!`

I found out that adding `lisitem.setInfo( type="video", infoLabels=None)` in `api.listitem()` in plex_api/base.py fixes the problem and the video resumes properly from this context menu item. However, this is probably not the best place to put that because it adds it to all listitems, but really we should only add it to listitems in the playlist, which I'm not sure where that is. 

4. If you add items to a playqueue that are partially played (or partially play them from a playqueue), and then you save the playqueue as a playlist .m3u, there are a couple unexpected things that happen. First, the .m3u file is saved like this:

```#EXTCPlayListM3U::M3U
#EXTINF:0,Roshi's Gambit
#EXT-KX-OFFSET:107000.0,0
plugin://plugin.video.plexkodiconnect.tvshows/499006/?plex_id=499017&plex_type=episode&mode=play&filename=Dragon Ball - S05E10 - Roshi's Gambit.mkv
#EXTINF:0,King Piccolo's Wish
plugin://plugin.video.plexkodiconnect.tvshows/499006/?plex_id=499018&plex_type=episode&mode=play&filename=Dragon Ball - S05E11 - King Piccolo's Wish.mkv```

You can see how in the partially played item (the first one), there is a "#EXT-KX-OFFSET:107000,0" which seems like a resume offset. We can ignore that (because Kodi handles resume) except for it breaks some things. Mainly: (1) Kodi thinks this item is a song and gives it a type of song instead of unknown in the playlist which breaks your if condition in _playback_triage to figure out if it's a playlist playback and doesn't play the item, and (2) the _m3u_iterator in playlists.db.py doesn't handle that extra line so it's giving an error on startup sync. 

For (1), see my suggested changed after this post. For (2), you need to change _m3u_iterator to handle 
 (ignore) the EXT-KX-OFFSET line by doing something like this

```def _m3u_iterator(text):
    """
    Yields e.g. plugin://plugin.video.plexkodiconnect.movies/?plex_id=xxx
    """
    lines = iter(text.split('\n'))
    for line in lines:
        if line.startswith('#EXTINF:'):
            next_line = next(lines).strip()
            if next_line.startswith('#EXT-KX-OFFSET:'):
                yield next(lines).strip()
            else:
                yield next_line```

**Comment by peacey** [comment]
Oh and just FYI, none of these problems with playlists happen if you use smart playlists. All the fanart shows up, buttons show up properly in context menu, nothing crashes, etc. You can easily construct a smart playlist that has certain episodes/movies ("File contains xx.mkv") so programming it would be simple. However, you cannot mix movies and tv show episodes in a smart playlist, which kind of defeats the purpose since Plex has playlists that mix that content and you sync it with Kodi. Currently the mixed content works fine in the m3u playlists . That's too bad for smart playlists, they would have been a really nice and elegant solution if Kodi allowed for mixed content in them.

**Comment by croneter** [comment]
Thanks again - your officially a PKC but hunter know 😆 . 

* I could not get Kodi to safe a resume point for a m3u playlist. How did you accomplish that? How did you navigate within Kodi?
* I also could not get Kodi to think that an episode is a song. How did you do that?

I'll split these things of in separate pull requests.

**Comment by peacey** [comment]
Lol, I've never used playlists before and when I started testing them I saw all these issues, and I was like dear god.

Btw, are you going to commit the offset change? You resolved the conversation but the changes haven't been committed for that suggestion. Are you adding it in another commit?
